src/main/java/com/google/devtools/build/lib/query2/output/JsonOutputFormatter.java

```diff
@@ -0,0 +1,105 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.devtools.build.lib.query2.output;
+
+import com.google.common.base.Preconditions;
+import com.google.devtools.build.lib.packages.Attribute;
+import com.google.devtools.build.lib.packages.Rule;
+import com.google.devtools.build.lib.packages.Target;
+import com.google.devtools.build.lib.query2.CommonQueryOptions;
+import com.google.devtools.build.lib.query2.engine.OutputFormatterCallback;
+import com.google.devtools.build.lib.query2.engine.QueryEnvironment;
+import com.google.devtools.build.lib.query2.engine.SynchronizedDelegatingOutputFormatterCallback;
+import com.google.devtools.build.lib.query2.engine.ThreadSafeOutputFormatterCallback;
+import com.google.devtools.build.lib.query2.output.OutputFormatter.AbstractUnorderedFormatter;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonObject;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+/** An output formatter that prints the result as Json. */
+class JsonOutputFormatter extends AbstractUnorderedFormatter {
+
+  @Override
+  public String getName() {
+    return "json";
+  }
+
+  @Override
+  public ThreadSafeOutputFormatterCallback<Target> createStreamCallback(
+      OutputStream out, QueryOptions options, QueryEnvironment<?> env) {
+    return new SynchronizedDelegatingOutputFormatterCallback<>(
+        createPostFactoStreamCallback(out, options));
+  }
+
+  @Override
+  public void setOptions(CommonQueryOptions options, AspectResolver aspectResolver) {
+    super.setOptions(options, aspectResolver);
+
+    Preconditions.checkArgument(options instanceof QueryOptions);
+  }
+
+  @Override
+  public OutputFormatterCallback<Target> createPostFactoStreamCallback(
+      final OutputStream out, final QueryOptions options) {
+    return new OutputFormatterCallback<Target>() {
+
+      private JsonObject result = new JsonObject();
+      private Gson gson = new GsonBuilder().setPrettyPrinting().create();
+
+      @Override
+      public void start() {}
+
+      @Override
+      public void processOutput(Iterable<Target> partialResult)
+          throws IOException, InterruptedException {
+        Set<String> hash_Set = new HashSet<>();
+        for (Target target : partialResult) {
+          result.add(target.getLabel().toString(), createTargetJsonObject(target, hash_Set));
+        }
+      }
+
+      @Override
+      public void close(boolean failFast) throws IOException {
+        if (!failFast) {
+          out.write(gson.toJson(result).getBytes());
+        }
+      }
+    };
+  }
+
+  private static JsonObject createTargetJsonObject(Target target, Set<String> hash_Set) {
+    JsonObject result = new JsonObject();
+    Gson gson = new Gson();
+    if (target instanceof Rule) {
```

Comment:
What if target is not instance of Rule? Should we handle those too?

Reply:
yep files etc. I should examine and handle those as well

Reply:
How about adding them in a list? Something like 
```
"//the:target": {
    "bazel.files": [
         <the list of files>
    ],
    ...
}
```

Reply:
since it is the target itself not a rule. the other options as I see are: ["PackageGroup", "OutputFile", "InputFile", "EnvironmentGroup"] which I haven't encountered yet and not sure what they are for. We can just replicate the behaviour in XmlFormatter but how about adding them in another pr ?

Reply:
do you know what are those?

Reply:
Hard to tell from the name only, especially `EnvironmentGroup` is cryptic. Let's leave it for later.